centrallix/expression/exp_functions.c

```diff
@@ -3935,146 +3936,241 @@ int exp_fn_lev_compare(pExpression tree, pParamObjects objlist, pExpression i0,
-			char *loc = strchr(CHAR_SET, toupper(term[i]));
-			if (loc) {
-				int index = (int)(loc - CHAR_SET);
-				table[index]++;
-			}
-		}
+	if(!ispunct(src[i]) && !isspace(src[i]))
+	    {
+	    dest[j] = tolower(src[i]);
+	    j++;
+	    }
+    	}
+    	return 0;
+    }
 
-		return 0;
-	}
 
-int exp_fn_i_relative_frequency_table(double *frequency_table)
-	{
-		int i;
-		double sum = 0;
-		// Compute the total character frequency
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			sum += frequency_table[i];
-		}
+/*
+ * hash_char_pair
+ * This method creates an vector table index based a given character pair. The characters are represented 
+ * as their ASCII code points.
+ *
+ * Parameters:
+ * 	num1 : first ASCII code point (double)
+ * 	num2 : second ASCII code point (double)
+ *
+ * Returns:
+ * 	vector table index (integer)
+ */
+int hash_char_pair(double num1, double num2)
+    {
+    int func_result = round(((num1 * num1 * num1) + (num2 * num2 * num2)) * ((num1+1)/(num2+1))) -1;
+    return func_result % VECTOR_TABLE_SIZE;	
+    }
 
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			frequency_table[i] = frequency_table[i] / sum;
-		}
-		return 0;
-	}
 
-int exp_fn_i_tf_idf_table(double *frequency_table)
+/*
+ * exp_fn_i_frequency_table
+ * This method creates a vector frequency table based on a string of characters.
+ *
+ * Parameters:
+ * 	table : integer pointer to vector frequency table
+ * 	term : the string of characters
+ *
+ * Returns:
+ * 	0 	
+ */
+int exp_fn_i_frequency_table(unsigned char *table, char *term)
+    {
+    int i;
+    // Initialize hash table with 0 values
+    for (i = 0; i < VECTOR_TABLE_SIZE; i++)
 	{
-		int i;
-		double sum = 0;
-		// Compute the total character frequency
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			sum += frequency_table[i];
-		}
-
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			frequency_table[i] = (frequency_table[i] / sum) * IDF[i];
-		}
-		return 0;
+	table[i] = 0;
 	}
+	// Trim out whitespace and punctuation and convert letters to lowercase
+    char* src = term;
+    char* dest = nmMalloc(strlen(term) + 1);
+    exp_fn_i_trim_term(src, dest);
+    term = dest;	
 
-// Dot product is equal to the sum of the squared values from each relative frequency table
-int exp_fn_i_dot_product(double *dot_product, double *r_freq_table1, double *r_freq_table2)
+    for(i = 0; i < strlen(term) + 1; i++)
 	{
-		int i;
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			*dot_product = *dot_product + (r_freq_table1[i] * r_freq_table2[i]);
+	// This case is for the first character. It adds a leading null character
+	if (i == 0)
+	    {
+	    double temp = (int)term[i];
+	    if (temp >= 48 && temp <= 57)
+		{
+		temp += 75;
+		} 
+					
+	    int index = hash_char_pair(96, temp);
+	    table[index] += ((unsigned char)temp + 96) % 13;
+	    }
+	    // This case is for characters in the middle of the string. It adds no null character
+	    else if (i < strlen(term))
+		{
+		double temp1 = (int)term[i-1];
```

Comment:
Assuming you do eliminate the string copy (see other comment), you'll need a way of referencing the previously processed character.  I'd suggest just a temp variable storing what the previous char was, or your code for NUL if there was no previous char.